### ISSUE TITLE
[ECS] [CodeDeploy] [CloudFormation]: CloudFormation support for BLUE/GREEN deployments on ECS

### ISSUE THREAD

**Issue by ilyasotkov** [aws/containers-roadmap#130]
The feature was announced  in November: https://www.youtube.com/watch?v=01ewawuL-IY

> For blue/green deployments, AWS CloudFormation supports deployments on AWS Lambda compute platforms only.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codedeploy-deploymentgroup.html

Not exactly sure whether [`AWS::ECS::Service.DeploymentConfiguration`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-deploymentconfiguration.html) would be affected as well.

**Comment by steven-cuthill-otm** [comment]
would be good to get more info on this and what are the plans.

**Comment by ngamradt-turner** [comment]
We would like to start experimenting with this feature, but without CloudFormation support, it is not practical for us to use.

**Comment by steven-cuthill-otm** [comment]
> We would like to start experimenting with this feature, but without CloudFormation support, it is not practical for us to use.

we have been promised that CFN will be getting 1st class treatment going forward ... we shall see :)

**Comment by deleugpn** [comment]
@steven-cuthill-otm who promised that?

**Comment by steven-cuthill-otm** [comment]
> @steven-cuthill-otm who promised that?
jeff barr, have a read ...
https://www.reddit.com/r/aws/comments/avx449/a_quick_cloudformation_update/

**Comment by deleugpn** [comment]
I read that thread and didn't see any moment the promise to get CloudFormation support for resources as 1st class treatment. The wording says they'll open their roadmap to allow us to tell what we would like to prioritize, which for me does not translate as getting 1st class treatment.

**Comment by sarojgharat** [comment]
Any update on this ?

**Comment by mmoulton** [comment]
I'm also very interested in this capability. Specifically I'd like to use it with CDK (https://github.com/awslabs/aws-cdk/issues/2056)

**Comment by angusfz** [comment]
Waiting for CFN support

**Comment by nitzanyemal123** [comment]
+ 1 For CF support

**Comment by andrewhiles** [comment]
Do we know if there's been any movement with this?

**Comment by JoeAlamo** [comment]
This would be a killer feature to have.

**Comment by paddie** [comment]
Yep, blocked by this as well. I think that this [article](https://frontier.town/2019/01/codedeploy-blue-green-deployments-and-cloudformation/) summarises pretty clearly why this is not a trivial thing for CF to handle:

> It turns out that there’s actually a very good reason why this is not a supported configuration. CodeDeploy is going to be referring to an Auto Scaling Group, and it’s likely that you have defined that ASG in your CloudFormation template, and used !Ref to link the deployment group to it.
>
> The problem comes when you do your first blue-green deployment. CodeDeploy does a blue-green deployment by cloning the ASG to make a new one with the same parameters, but whose instances run the newer version of your application. Once the new version is deployed and healthy, it winds down the old ASG. Then, it updates the deployment group configuration to point to the new ASG, and deletes the old ASG.
>
> This is fine until you go back to your CloudFormation template and need to make a change that touches the ASG. CloudFormation will try to update the ASG that it created, only to find it doesn’t exist any more. Boom, template update failure and rollback.

**Comment by ScOut3R** [comment]
In regards to ECS blue/green deployments CodeDeploy uses two already existing Target Groups to switch traffic back and forth. While the referenced article holds correct information it does not relate to ECS blue/green deployments but EC2 based ones.
I have a service where I deployed almost everything via CloudFormation, including the two Target Groups used for blue/green deployments. I provisioned the CodeDeploy side with Terraform and everything works quite neatly. Since CodeDeploy does not destroy/clone/recreate the Target Groups both parties are happy.
Apart from capacity concerns as a consumer I don't see any blockers for ECS support.

**Comment by okram999** [comment]
@ScOut3R - how do the service know that the deployment type is bluegreen, if you configure them using cloudformation. The option to specify `deployment type` as bluegreen is what where the challenge starts. no?

**Comment by ScOut3R** [comment]
@okram999 The blue/green CodeDeploy configuration is deployed by Terraform. I pass in as input variables the ECS Service and the two Target Groups (which are built by CloudForamtion) to Terraform which builds the CodeDeploy and CodePipeline configuration.

**Comment by noahxp** [comment]
+1 For CF support

**Comment by pferretti** [comment]
+1 For CF support

**Comment by ngamradt-turner** [comment]
Is there an update on this feature?

**Comment by mwarkentin** [comment]
Probably helpful to add your votes and follow on the issues on the cloudformation roadmap:

* https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/56
* https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/37

**Comment by dslove** [comment]
I need this too.

**Comment by ramanasatyavolu** [comment]
Is there a feature provided in CloudFormation to launch blue green deployment through ECS

**Comment by surajtikoo** [comment]
Waiting for a long time .Is any time line when it be be available in cloudFormation to launch  blue green deployment through ecs fargare?

**Comment by mwarkentin** [comment]
Looks like this just showed up: https://docs.aws.amazon.com/en_pv/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-deploymentcontroller.html

Haven’t had a chance to try it.

**Comment by surajtikoo** [comment]
@mwarkentin  Thanks for your Update, but i think with the new feature of ALB which support multiple target group with  single Listeners is quite amazing. We can use weighted routing which is much easier and faster. 
You are quite late with this feature. :)

https://aws.amazon.com/blogs/aws/new-advanced-request-routing-for-aws-application-load-balancers/

**Comment by deleugpn** [comment]
I don't see how the ALB weighted request fit into CI/CD strategy. I want a limited time for weighted request instead of supporting 2 versions at all times.

**Comment by surajtikoo** [comment]
@deleugpn. You can still handle this with your custom pipeline in build step to route the traffic, but i agree with your point if we need to test before forwarding traffic to your live environment then  multiple  versions support is quite useful.

**Comment by ngamradt-turner** [comment]
Since this is still in the "Coming Soon" column, I am guessing that this feature will not come out until after re:Invent, or will something be announced yet during re:Invent?

**Comment by deleugpn** [comment]
I think by this point it's crystal clear that AWS so not disclose release date until the day of the release.

**Comment by cb-manideep** [comment]
+1 Need this.

**Comment by Lasim** [comment]
+1

**Comment by Chidiebube** [comment]
Hi @coultn  I see that there a property value CODE_DEPLOY in the "DeploymentController:" property docs [here ](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-service-deploymentcontroller.html) does this mean that this is now available ? And this issue need to be updated ?

**Comment by mwarkentin** [comment]
Also hoping that this will work with the new CodeDeploy canary deployment strategies!

https://github.com/aws/containers-roadmap/issues/229

**Comment by richardmaltais** [comment]
@paddie For that specific matter, using Blue-Green with CodeDeploy on EC2 instances, we used a custom setup to keep the original ASG, then CodeDeploy-generated ASGs are switched over as usual, drained when needed and the updated LaunchConfiguration is also "copied" over through a Lambda, so that CloudFormation doesn't freak out when comes the time to update it. However, it comes with a few limitations:

- not possible to change the capacity of the ASG (the CodeDeploy-generated one)
- we have to have set the capacity to 0 when applying changes to the LaunchConfiguration and the ASG. Then, the Lambdas do the rest

However, we didn't try to experiment with resource import for the CodeDeploy generated ASG.

**Comment by tuttlem** [comment]
+1 For CF support

**Comment by KiamarzFallahi** [comment]
We are excited to announce you can now use AWS CloudFormation to perform Amazon ECS blue/green and canary deployments through AWS CodeDeploy. Blue/green deployments are a safe deployment strategy provided by AWS CodeDeploy for minimizing interruptions caused by changing application versions.

To learn more visit [our announcement](https://aws.amazon.com/about-aws/whats-new/2020/05/aws-cloudformation-now-supports-blue-green-deployments-for-amazon-ecs/) and the [user guide](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/blue-green.html).

**Comment by sethstone** [comment]
I've been tracking this ticket for a few months and I was excited to see it close, however, the solution provided did not solve the problem I expected.  It seems that the new capability allows one to use Cloudformation as the deployment controller - in particular you would be monitoring the deployment in Cloudformation rather than the CodeDeploy console and release changes through stack updates.  This may be exactly what others were looking for, but I was expecting was to have the ability to create a deployment group (with CF) using the ECS BlueGreen deployment type and be able to manage the deployment process (once created by CF) in the CodeDeploy console. 

Here's an example that I hoped would work after this change

```
 DeploymentGroup:
    Type: AWS::CodeDeploy::DeploymentGroup
    Properties: 
      ApplicationName: !Ref CodeDeployApplication
      AutoRollbackConfiguration: 
        Enabled: true
        Events:
          - DEPLOYMENT_FAILURE
      BlueGreenDeploymentConfiguration: 
        DeploymentReadyOption: 
          ActionOnTimeout: CONTINUE_DEPLOYMENT
          WaitTimeInMinutes: 0
        TerminateBlueInstancesOnDeploymentSuccess:
          Action: TERMINATE
          TerminationWaitTimeInMinutes: 5
      DeploymentConfigName: CodeDeployDefault.ECSAllAtOnce
      DeploymentGroupName: !Ref CodeDeployDeploymentGroupName
      DeploymentStyle: 
        DeploymentOption: WITH_TRAFFIC_CONTROL
        DeploymentType: BLUE_GREEN
      LoadBalancerInfo: 
        TargetGroupPairInfoList: 
          TargetGroups:
            - Name: !Sub ${ProjectPrefix}-tg1
            - Name: !Sub ${ProjectPrefix}-tg2
          ProdTrafficRoute: 
            ListenerArns: !Ref ALBProductionListener
      ServiceRoleArn: !GetAtt CodeDeployServiceRole.Arn
      ECSServices:
        - ServiceName: !Sub ${ProjectPrefix}-service
        - ClusterName: !Sub ${ProjectPrefix}-ecs-fargate-cluster
```

Based on the Youtube video I thought this is what we were trying to accomplish because it shows several Deploy stages with ECS Blue/Green Deploy (not a Cloudformation deploy).  

Could someone let me know if this is an issue that is being worked on on the roadmap and/or am I doing something wrong to expect this to work?  (I based this CF template on a working CLI command that I have)

Thanks!

**Comment by KiamarzFallahi** [comment]
@sethstone You are correct. This feature does not support creating and configuring a CodeDeploy deployment group to naively perform blue/green deployments for ECS. That is a coverage gap being tracked: https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/37

**Comment by clareliguori** [comment]
That issue was also closed:
https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/37#issuecomment-632784182

@sethstone A new issue was created on the CloudFormation roadmap here requesting coverage in `AWS::CodeDeploy::DeploymentGroup` for ECS blue-green, it would be great for you to comment there with your example template snippet. Thanks!
https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/483

**Comment by mwarkentin** [comment]
There are a couple of limitations which mean that this feature won't work with our ECS platform ([convox](https://convox.com/)):

* Declaring output values or importing values from other stacks is not currently supported for templates defining blue/green ECS deployments. 
* You cannot use the AWS::CodeDeploy::BlueGreen hook in a template that includes nested stack resources.
* You cannot use the AWS::CodeDeploy::BlueGreen hook in a nested stack.

Are there plans / work in progress to handle these use cases? Or for more complex use cases like this would it be meant to be handled by the CodeDeploy DeploymentGroup support described above?

**Comment by Lasim** [comment]
Thanks for the analysis @mwarkentin.

it's meant to me that there is no sense yet to use this feature if these functions are not implemented.

**Comment by vinay-nadig-0042** [comment]
I see that the issue has been closed since the Cloudformation support has been announced. But, as @mwarketin pointed it out, the limitations are quite crippling. Is there a separate issue where these limitations and their fixes are tracked?

**Comment by zzenonn** [comment]
> I see that the issue has been closed since the Cloudformation support has been announced. But, as @mwarketin pointed it out, the limitations are quite crippling. Is there a separate issue where these limitations and their fixes are tracked?

I opened this issue on the CloudFormation roadmap

https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/483

**Comment by simi-obs** [comment]
June 24th 2022.

Does anybody know how to control blue/green ECS deployment using Cloudformation(CF) without all the stupid limitations that the hook imposes (no outputs, no dynamic references, no other updates ...)?

I mean at this point I am OK with implementing my own custom resource or even custom resource type. I am willing to bend and punish CF in ugly ways, the only thing I want is a way to perform ECS blue/green deployment as a part of CF deployment (update) process without all the stupid limits.

**Comment by teekennedy** [comment]
@simi-obs I can assure you that using a custom resource as a workaround for this is not a dead end. I implemented one myself (my very first lambda actually) back in 2019 when I started watching this issue and it's been used thousands of times over in production with a high success rate.

I unfortunately can't post or give you the code without express permission from my employer, but I'll try to give you the rundown of the overall process and the gotchas / lessons learned along the way.

In 2019, it was not possible to create a CloudFormation custom resource that was responsible for just the deployment itself (the "control resource" as you call it), because some the configuration required to create a CodeDeploy enabled ECS service was not allowed by the existing `AWS::ECS::Service` resource type. I ended up replacing the ECS service resource itself with my custom resource lambda. In addition to managing the ECS service, it managed the CodeDeploy application, and deployment group resources. Whether that was required or just made my life easier I can't recall specifically anymore.

I suggest you approach writing your lambda on an event-by-event basis:
- The _create_ and _delete_ events are likely no-ops unless you end up managing other resources.
- The _update_ event should:
  - Use some kind of locking or retry mechanism to avoid creating a new deployment while one is already in progress. I ended up enforcing that all deployments in the CI/CD pipeline go through CloudFormation to try to prevent this situation, and then wrote a simple polling loop to wait for any in-progress deployments to finish before initiating a new one as a fallback measure. There are better ways, but this solution has yet to cause issues so it is still in use.
  - Collect the parameters for the deployment from the properties of the resource, including at least partially generating the app spec for the deployment.
  - Create the deployment, then wait for it to enter the `Failed` or `Succeeded` status.
  - Return the corresponding error or success signal to CloudFormation.

The caveats to watch out for:
- Make sure you send a response to CloudFormation in every scenario. Not sending a response means waiting for CloudFormation to time out against the custom resource, which IIRC defaults to 4 hours. If you end up in this situation, there's nothing you can do but wait, so do everything you can to avoid it. Catch any exceptions and return an error response. Send an error response if the lambda is about to time out. Carefully review the logic used to form and send the response to ensure it always sends something.
- As mentioned above, make sure you handle the use case of overlapping deployments gracefully. The details of this depend on how the deployments fit into the overall release process.
- Deployments can run for much longer than lambdas. There are some CloudFormation custom resource libraries that chain multiple lambda executions together using continuation passing lambda events to work around this. You could also stop the in-progress deployment if the lambda is getting too close to timing out.

Hope this helps!

**Comment by simi-obs** [comment]
@teekennedy Very nice breakdown, thank you.
Some of the caveats you mentioned have solutions nowadays:
- I.e. instead of using the Lambda backed custom resource, I will implement Cloudformation [custom resource type](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/registry.html), which does not have some of the limitations (done that in the past).

Regarding the overlapping deployments: 
- We are using isolated environments, which are updated only using Cloudformation. What I intend to do is to make sure(enforce) that if the Cloudformation stack update is not ongoing, NO code deploy deployment (of the stack's services) is ongoing.  This is in line with what you have mentioned: I will make sure I always wait for the code deploy deployment to either Fail or Succeed. 
If I can enforce this, I should be able to enforce no overlapping deployments (since by default the cannot be overlapping Cloudformation stack updates).

Also I believe I will still have to handle the ECS:Service as a part of the custom resource. Otherwise, I might run into the problem with Cloudformation trying to replace the service (which would probably end up in error) once the TaskDefinition changes.

Overall, thank you for all the tips, will let the community here know how I ended up.
